cirq/ops/linear_operator.py

```diff
@@ -0,0 +1,316 @@
+    @abc.abstractmethod
+    def _n_dim_(self) -> Optional[int]:
+        """Returns the number of dimensions of the space self acts on."""
+        pass
+
+    @abc.abstractmethod
+    def _matrix_(self) -> Optional[np.ndarray]:
+        """Returns the matrix representation of self."""
+        pass
+
+    @abc.abstractmethod
+    def _pauli_expansion_(self) -> Optional[np.ndarray]:
+        """Returns Pauli basis expansion of self.
+
+        This representation is currently only used for single-qubit operators.
+        """
+        pass
+
+    def matrix(self) -> Optional[np.ndarray]:
+        """Returns or computes the matrix corresponding to self."""
+        matrix = self._matrix_()
+        if matrix is not None:
+            return matrix
+
+        pauli_expansion = self._pauli_expansion_()
+        if pauli_expansion is not None:
+            return operator_spaces.reconstruct_from_expansion(
+                pauli_expansion, operator_spaces.PAULI_BASIS)
+
+        return None
+
+    def pauli_expansion(self) -> Optional[np.ndarray]:
+        """Returns or computes expansion of self in the Pauli basis."""
+        if self._n_dim_() != 2:
+            return None
+
+        pauli_expansion = self._pauli_expansion_()
+        if pauli_expansion is not None:
+            return pauli_expansion
+
+        matrix = self._matrix_()
+        if matrix is not None:
+            return operator_spaces.expand_in_basis(
+                matrix, operator_spaces.PAULI_BASIS)
+
+        return None
+
+    def __add__(
+            self,
+            other: 'AbstractLinearOperator'
+    ) -> 'AbstractLinearOperator':
+        """Computes vector addition in the space of linear operators."""
+        pauli_expansion = None
+        pauli_expansion_a = self._pauli_expansion_()
+        pauli_expansion_b = other._pauli_expansion_()
+        if pauli_expansion_a is not None and pauli_expansion_b is not None:
+            pauli_expansion = pauli_expansion_a + pauli_expansion_b
+
+        matrix = None
+        matrix_a = self.matrix()
+        matrix_b = other.matrix()
+        if matrix_a is not None and matrix_b is not None:
+            matrix = matrix_a + matrix_b
+
+        return LinearOperator(matrix, pauli_expansion)
+
+    def __sub__(
+            self,
+            other: 'AbstractLinearOperator'
+    ) -> 'AbstractLinearOperator':
+        """Computes difference between self and other."""
+        return self.__add__(-other)
+
+    def __neg__(self) -> 'AbstractLinearOperator':
+        """Computes additive inverse of self."""
+        return self.__mul__(-1)
+
+    def __mul__(
+            self,
+            other: Union[complex, float, int]
+    ) -> 'AbstractLinearOperator':
+        """Computes scalar multiplication in the space of linear operators."""
+        pauli_expansion = self._pauli_expansion_()
+        if pauli_expansion is not None:
+            pauli_expansion = pauli_expansion * other
+
+        matrix = self._matrix_()
```

Comment:
Same comment as addition: I don't think we want to be computing the entire matrix for basic operations.

Reply:
`_matrix_()` doesn't compute the matrix if one isn't stored in the operator.